feat: add `OptionsViewOpenDirection.mostSpace` to `RawAutocomplete`

SalehTZ opened the pull request:
This PR changes the default value of `RawAutocomplete.optionsViewOpenDirection` from `OptionsViewOpenDirection.down` to `OptionsViewOpenDirection.mostSpace`.

When `optionsViewOpenDirection` is set to `mostSpace`, the options view will open in the direction with the most available space. This behavior is determined by comparing the space above and below the text field.

The commit also updates the logic for calculating `optionsViewMaxHeight` and `originY` to handle the new `mostSpace` option, and ensures that the `switch` statement for `alignment` correctly handles this new case.


https://github.com/user-attachments/assets/8e00318a-bd07-402b-b516-543740cd5c26



fixes #101620 


## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

**Note**: The Flutter team is currently trialing the use of [Gemini Code Assist for GitHub](https://developers.google.com/gemini-code-assist/docs/review-github-code). Comments from the `gemini-code-assist` bot should not be taken as authoritative feedback from the Flutter team. If you find its comments useful you can update your code accordingly, but if you are unsure or disagree with the feedback, please feel free to wait for a Flutter team member's review for guidance on which automated comments should be addressed.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

SalehTZ committed:
Make RawAutocomplete.optionsViewOpenDirection default to 'auto'

This commit changes the default value of `RawAutocomplete.optionsViewOpenDirection` from `OptionsViewOpenDirection.down` to `OptionsViewOpenDirection.auto`.

When `optionsViewOpenDirection` is set to `auto`, the options view will open in the direction with the most available space. This behavior is determined by comparing the space above and below the text field.

The commit also updates the logic for calculating `optionsViewMaxHeight` and `originY` to handle the new `auto` option, and ensures that the `switch` statement for `alignment` correctly handles this new case.

SalehTZ committed:
Add 'auto' option to Autocomplete OptionsViewOpenDirection

This commit introduces an 'auto' option for `OptionsViewOpenDirection` in the Material `Autocomplete` widget.

When `OptionsViewOpenDirection.auto` is selected, the options view will open downwards in landscape orientation and upwards in portrait orientation.

The default value for `optionsViewOpenDirection` has been changed from `down` to `auto`.

SalehTZ committed:
[autocomplete] Add tests for OptionsViewOpenDirection.auto

This commit adds tests to verify the behavior of `OptionsViewOpenDirection.auto` in the `RawAutocomplete` widget.

The tests ensure that when `OptionsViewOpenDirection.auto` is used:
- The options view opens downwards if there is enough space below the text field.
- The options view opens upwards if there is not enough space below but enough space above the text field.

Additionally, redundant argument values for `optionsViewOpenDirection` were removed in existing tests.

SalehTZ committed:
Add test for RawAutocomplete with OptionsViewOpenDirection.auto

This test verifies that the options view opens in the direction with more available space when `optionsViewOpenDirection` is set to `OptionsViewOpenDirection.auto`.

It first positions the Autocomplete widget at the top and expects the options to open downwards. Then, it moves the widget to the bottom and expects the options to open upwards.

SalehTZ committed:
Fix Autocomplete optionsViewOpenDirection default value

The default value for `optionsViewOpenDirection` in `Autocomplete` was incorrectly documented and tested as `OptionsViewOpenDirection.down`. This commit corrects the test to reflect the actual default value, which is `OptionsViewOpenDirection.auto`.

SalehTZ committed:
Test `OptionsViewOpenDirection.auto` in Autocomplete

This commit adds a test to verify that when `optionsViewOpenDirection` is set to `OptionsViewOpenDirection.auto`, the options view opens in the direction where there is more available space.

SalehTZ committed:
fix test failure due to `avoid_redundant_argument_values` lint

SalehTZ committed:
fix options alignment on `auto` case

SalehTZ committed:
hmm incorrect comment

victorsanni commented on packages/flutter/lib/src/material/autocomplete.dart:
```diff
@@ -67,7 +67,7 @@ class Autocomplete<T extends Object> extends StatelessWidget {
     this.optionsMaxHeight = 200.0,
     this.optionsViewBuilder,
-    this.optionsViewOpenDirection = OptionsViewOpenDirection.down,
+    this.optionsViewOpenDirection = OptionsViewOpenDirection.auto,
```
Changing the default value is a breaking change, can we keep the previous default and just add the new option?

SalehTZ replied:
Yes sir, I agree.

victorsanni commented on packages/flutter/lib/src/material/autocomplete.dart:
```diff
@@ -211,6 +211,8 @@ class _AutocompleteOptions<T extends Object> extends StatelessWidget {
     final AlignmentDirectional optionsAlignment = switch (openDirection) {
       OptionsViewOpenDirection.up => AlignmentDirectional.bottomStart,
       OptionsViewOpenDirection.down => AlignmentDirectional.topStart,
+      // defaults to down
```
Not sure this is needed.

victorsanni commented on packages/flutter/lib/src/widgets/autocomplete.dart:
```diff
@@ -106,6 +106,9 @@ enum OptionsViewOpenDirection {
   down,
+
+  /// Open in the direction with the most available space.
+  auto,
```
I think we should be explicit and just use `automatic` here. This is consistent with other similar properties in the framework as well (`AutomaticKeepAlive`, `automaticallyImplyLeading`, `NavigationBarBottomMode.automatic` etc.

SalehTZ replied:
sure, done.

victorsanni commented on packages/flutter/lib/src/widgets/autocomplete.dart:
```diff
@@ -106,6 +106,9 @@ enum OptionsViewOpenDirection {
   /// of the text field built by [RawAutocomplete.fieldViewBuilder].
   down,
+
+  /// Open in the direction with the most available space.
```
What happens when both directions have the same available space? This should be called out in the documentation.

SalehTZ replied:
When `spaceAbove` and `spaceBelow` are equal, `OptionsViewOpenDirection.down` will be selected.

Good point, I'll mention this.

SalehTZ committed:
omitted comment

SalehTZ committed:
renamed `auto` to `automatic`.

SalehTZ committed:
commented on unreachable lines

SalehTZ committed:
roll back `optionsViewOpenDirection` to `OptionsViewOpenDirection.down`

SalehTZ committed:
rollback default `optionsViewOpenDirection` to `OptionsViewOpenDirection.down`

SalehTZ committed:
rollback default `optionsViewOpenDirection` to `OptionsViewOpenDirection.down`

Also added a test to test when field alignment is center, in which direction options menu open

SalehTZ committed:
docs: `automatic` of `OptionsViewOpenDirection` opens downwards if same space

If both directions have the same available space, then `OptionsViewOpenDirection.down` will be selected.

SalehTZ committed:
Merge remote-tracking branch 'origin/feat/add-OptionsViewOpenDirection.auto' into feat/add-OptionsViewOpenDirection.auto

# Conflicts:
#	packages/flutter/lib/src/widgets/autocomplete.dart

SalehTZ committed:
Merge branch 'master' into feat/add-OptionsViewOpenDirection.auto

victorsanni commented on packages/flutter/lib/src/widgets/autocomplete.dart:
```diff
@@ -106,6 +106,12 @@ enum OptionsViewOpenDirection {
+  /// Open in the direction with the most available space.
+  ///
+  /// If both directions have the same available space,
+  /// then `OptionsViewOpenDirection.down` will be selected.
```
Nit:

```suggestion
  /// the options view opens downward.
```

victorsanni reviewed (approved):
LGTM, thanks for the PR.

justinmc commented on packages/flutter/lib/src/widgets/autocomplete.dart:
```diff
@@ -527,19 +533,30 @@ class _RawAutocompleteState<T extends Object> extends State<RawAutocomplete<T>>
+    final double optionsViewMaxHeight = switch (optionsViewOpenDirection) {
       OptionsViewOpenDirection.up => -overlayRectInField.top,
       OptionsViewOpenDirection.down => overlayRectInField.bottom - fieldSize.height,
+      OptionsViewOpenDirection.automatic => 0.0, // Should be unreachable.
```
Nit: To avoid the need for these, what if you replaced the local optionsViewOpenDirection variable with a boolean opensUp?

SalehTZ replied:
hmm, maybe. That's better.

justinmc reviewed (commented):
At first I was questioning your approach of placing the options on the side with the most space, but after some thought I think you're probably right... 

Context menus do something slightly different. They try to position themselves above the content, and if they don't fit, then they position themselves below (see [TextSelectionToolbarLayoutDelegate](https://main-api.flutter.dev/flutter/widgets/TextSelectionToolbarLayoutDelegate-class.html)). Maybe this approach doesn't make sense for Autocomplete because the size of the options changes? It would be strange for the options to flip from above to below as the user types and gets different numbers of options.

However, what about the case where the Autocomplete is in a vertical scrollable? I guess if you scroll while the options are visible then they will flip between top and bottom at the middle of the screen. Maybe this is what they should do, though.

SalehTZ committed:
nit

Co-authored-by: Victor Sanni <victorsanniay@gmail.com>

SalehTZ committed:
Nit: introduced `opensUp` variable for better code style.

SalehTZ committed:
Merge branch 'feat/add-OptionsViewOpenDirection.auto' of github.com:SalehTZ/flutter into feat/add-OptionsViewOpenDirection.auto

SalehTZ committed:
fix: introduced opensUp variable.

justinmc commented on packages/flutter/lib/src/widgets/autocomplete.dart:
```diff
@@ -527,20 +533,25 @@ class _RawAutocompleteState<T extends Object> extends State<RawAutocomplete<T>>
-      OptionsViewOpenDirection.up => -overlayRectInField.top,
-      OptionsViewOpenDirection.down => overlayRectInField.bottom - fieldSize.height,
-    };
+    bool opensUp = widget.optionsViewOpenDirection == OptionsViewOpenDirection.up;
+    if (widget.optionsViewOpenDirection == OptionsViewOpenDirection.automatic) {
+      final double spaceAbove = -overlayRectInField.top;
+      final double spaceBelow = overlayRectInField.bottom - fieldSize.height;
+      opensUp = spaceAbove > spaceBelow;
+    }
```
Nit: Would this be more clear as a switch statement on widget.optionsViewOpenDirection? Then you could also make opensUp `final`.

SalehTZ replied:
Agree. done.

justinmc commented on packages/flutter/test/material/autocomplete_test.dart:
```diff
@@ -594,6 +594,91 @@ void main() {
+      setState(() {
+        alignment = Alignment.bottomCenter;
+      });
+      await tester.pumpAndSettle();
```
Could this be just a single pump?

justinmc commented on packages/flutter/test/material/autocomplete_test.dart:
```diff
@@ -594,6 +594,91 @@ void main() {
+      setState(() {
+        alignment = Alignment.center;
+      });
+      await tester.pumpAndSettle();
```
Same question here about a single pump.

justinmc commented on packages/flutter/test/widgets/autocomplete_test.dart:
```diff
@@ -984,6 +1006,76 @@ void main() {
+
+      // Show the options. It should open downwards since there is more space.
+      await tester.tap(find.byType(TextField));
+      await tester.pumpAndSettle();
```
And here and below about one pump.

justinmc reviewed (approved):
LGTM with nits 👍 . Thank you!

LongCatIsLooong reviewed (commented):
Haven't looked into the details but I have some concerns:

The "available space" in the implementation seems to refer to the space between the text field to the edges of the Overlay. This is can be different from user-perceived space as there may be other UI that will obstruct the options view (e.g., software keyboard, iOS dynamic island), and I think currently it's possible for the Overlay to be obstruct by things like the app bar (if it's not on the root Overlay). 

Also the hard coded threshold (0.5 * overlay height) seems a bit strange. This is probably a UX/design decision that should be made by the app developers, especially since we don't know if there's gonna be "floating" UI in the app that can get in the way of the options view (like floating action buttons).

SalehTZ committed:
nit: make `opensUp` final also use witch case instead of if statement

SalehTZ committed:
nit: use `pump` instead of `pumpAndSettle`

SalehTZ committed:
nit: use switch case for better code readability

SalehTZ commented on packages/flutter/lib/src/material/autocomplete.dart:
```diff
@@ -208,10 +208,10 @@ class _AutocompleteOptions<T extends Object> extends StatelessWidget {
-      OptionsViewOpenDirection.up => AlignmentDirectional.bottomStart,
-      OptionsViewOpenDirection.down => AlignmentDirectional.topStart,
-    };
+    final bool opensUp = openDirection == OptionsViewOpenDirection.up;
```
I'm thinking why do we even need this Align widget? 🤔

SalehTZ replied:
That's right.

* The overlay figures out how much space the options can use (`optionsViewBoundingBox`) and applies tight constraints to the child. It then uses a single `Align` to anchor the options either to the **top** (for downwards opening) or **bottom** (for upwards opening), so they always appear attached to the field.
* Having another `Align` inside Material’s `_AutocompleteOptions` is redundant. Since the parent box is usually bigger than the child, that inner `Align` can actually fight the outer one, which can lead to unexpected layouts.

Removing the inner `Align` makes the default Material-style Autocomplete behave predictably. It keeps all the alignment logic in one place—the overlay—so it’s easier to understand. And users can still provide their own `optionsViewBuilder` if they want full control.

LongCatIsLooong replied:
Great to hear! What's the solution you have in mind?

Also I re-read my comment, FABs should probably be obscured by option views so that was a good example. A good example would be in-app keyboards (custom keyboards that some apps add, as a part of the widget tree).

SalehTZ replied:
The first time the keyboard (IME) opens, the autocomplete options overlay may render underneath the keyboard.
This happens because `_buildOptionsView` calculates available space only once when the overlay is first built, and it isn’t automatically notified that `viewInsets` (keyboard height) has changed until some other rebuild occurs (like typing).

So we can make `_RawAutocompleteState` a `WidgetsBindingObserver` and override `didChangeMetrics()`.
`didChangeMetrics()` is triggered whenever the system window metrics change, including:
* when the keyboard opens/closes (changes window.viewInsets),
* when the device rotates,
* or when system UI insets change.

Inside `didChangeMetrics()`, we call `setState()`.
This forces a rebuild of the overlay, which in turn calls `_buildOptionsView` again.
On rebuild, the space above/below is recalculated using the new IME height (`window.viewInsets.bottom`), ensuring the options menu always adjusts immediately to the keyboard.

(Didn't test this 👇🏻 )
Alternatively, we can wrap the overlay child in a `MediaQuery` that passes down the latest `viewInsets`.
Since `MediaQuery` automatically updates whenever the keyboard opens or closes, this approach ensures the options view recalculates its size without needing an explicit observer. It’s also a bit lighter since it reuses Flutter’s existing rebuild mechanism instead of adding manual metric listeners.

SalehTZ committed:
removed `Align` widget.

SalehTZ committed:
Merge branch 'flutter:master' into feat/add-OptionsViewOpenDirection.auto

SalehTZ committed:
feat(autocomplete): add optionsViewOpenDirectionBuilder for flexible overlay direction control

Adds a new `optionsViewOpenDirectionBuilder` parameter to `RawAutocomplete` that
allows developers to customize how the options view's open direction is determined
based on available space.

Key changes:
- Added `OptionsViewOpenDirectionBuilder` typedef for the builder signature
- Added `optionsViewOpenDirectionBuilder` parameter to `RawAutocomplete`
- Updated logic to use the builder when provided, falling back to the existing
  `optionsViewOpenDirection` behavior
- Improved documentation to clarify how space is calculated within the Overlay

This change enables more flexible behaviors like:
- Setting minimum space requirements before opening in a direction
- Implementing different thresholds for up/down directions
- Creating custom logic that considers other factors beyond just space

The existing `optionsViewOpenDirection` enum is maintained for backward
compatibility, with the builder taking precedence when both are provided.

SalehTZ committed:
Merge branch 'flutter:master' into feat/add-OptionsViewOpenDirection.auto

SalehTZ committed:
fix: fix docs issue

SalehTZ committed:
Merge branch 'feat/add-OptionsViewOpenDirection.auto' of github.com:SalehTZ/flutter into feat/add-OptionsViewOpenDirection.auto

SalehTZ committed:
fix: fixed doc issue

SalehTZ committed:
fix: doc code snippet fix

SalehTZ committed:
feat(autocomplete): add optionsViewOpenDirectionBuilder for flexible overlay direction control

Adds a new `optionsViewOpenDirectionBuilder` parameter to both `RawAutocomplete`
and Material `Autocomplete` widgets, allowing developers to customize how the
options view's open direction is determined based on available space.

Key changes:
- Added `OptionsViewOpenDirectionBuilder` typedef
- Added `optionsViewOpenDirectionBuilder` to both widgets
- Updated documentation with examples
- Maintained backward compatibility with existing `optionsViewOpenDirection`

This enables more flexible behaviors like:
- Setting minimum space requirements
- Different thresholds for up/down directions
- Custom logic based on available space

Example usage:
```dart
optionsViewOpenDirectionBuilder: (double spaceAbove, double spaceBelow) {
  // Prefer opening up if there's enough space
  if (spaceAbove >= 200) return OptionsViewOpenDirection.up;
  // Otherwise open down if there's enough space
  if (spaceBelow >= 100) return OptionsViewOpenDirection.down;
  // Fallback to up if space is limited
  return OptionsViewOpenDirection.up;
}
```

SalehTZ committed:
Merge branch 'master' into feat/add-OptionsViewOpenDirection.auto

SalehTZ committed:
Merge branch 'master' into feat/add-OptionsViewOpenDirection.auto

dkwingsmt commented:
@LongCatIsLooong Just to confirm, is the discussion still going on with this PR? (from triage)

LongCatIsLooong commented on packages/flutter/lib/src/widgets/autocomplete.dart:
```diff
@@ -89,6 +89,11 @@ typedef AutocompleteOptionToString<T extends Object> = String Function(T option)
 /// A direction in which to open the options-view overlay.
 ///
+/// The available space for the options view is calculated based on the
+/// constraints of the [Overlay] that contains the autocomplete widget,
```
nit: constaints -> size

LongCatIsLooong replied:
Also this is only relevant to `automatic` so consider putting the docs there instead

LongCatIsLooong commented on packages/flutter/lib/src/widgets/autocomplete.dart:
```diff
@@ -108,8 +113,52 @@ enum OptionsViewOpenDirection {
+  /// Consider using [RawAutocomplete.optionsViewOpenDirectionBuilder] if you need more control
+  /// over the direction logic, such as minimum space requirements or different
+  /// thresholds for each direction.
+  automatic,
```
gemini suggested `.mostSpace` when I asked for a name that would make the underlying strategy clear.

SalehTZ replied:
Thanks good point. But I think sometimes it's not `most space`, it might be `less space`.

SalehTZ replied:
With latest commit, that would be good suggestion.

LongCatIsLooong commented on packages/flutter/lib/src/widgets/autocomplete.dart:
```diff
@@ -236,12 +286,34 @@ class RawAutocomplete<T extends Object> extends StatefulWidget {
+  ///  * [OptionsViewOpenDirection], for predefined open directions.
+  ///  * [OptionsViewOpenDirectionBuilder], for the signature of this builder.
+  /// {@endtemplate}
+  final OptionsViewOpenDirectionBuilder? optionsViewOpenDirectionBuilder;
```
Having 2 different parameters for determining whether the options view should appear is confusing. 

Instead, you can turn `OptionsViewOpenDirection` into a class with 3 static members: `.above`, `.below`, `.auto`.

SalehTZ replied:
I suggested that approach based on the [discussion](https://github.com/flutter/flutter/pull/172997#discussion_r2316812073) we had earlier.  
I agree that simplifying it to a single parameter makes sense. Sure, Let’s defer the custom logic to a separate PR so we can keep this one more focused.

Do I revert it?

LongCatIsLooong replied:
Yeah the rest of the PR LGTM so it will probably faster that we remove that new parameter and merge this one, and introduce that feature in a separate PR.

SalehTZ replied:
Done.

SalehTZ committed:
Merge branch 'master' into feat/add-OptionsViewOpenDirection.auto

SalehTZ committed:
nit: document correction

Signed-off-by: Alex Tz <31685655+SalehTZ@users.noreply.github.com>

SalehTZ committed:
removed `OptionsViewOpenDirectionBuilder` and rollback changes.

SalehTZ committed:
refactor: renamed `.automatic` option to `.mostSpace`

SalehTZ committed:
refactor: make docs less confusing

SalehTZ committed:
Merge branch 'master' into feat/add-OptionsViewOpenDirection.auto
